classes/secret_manager.php

```diff
@@ -0,0 +1,325 @@
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * MFA secret management class.
+ *
+ * @package     tool_mfa
+ * @author      Peter Burnett <peterburnett@catalyst-au.net>
+ * @copyright   Catalyst IT
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace tool_mfa;
+
+defined('MOODLE_INTERNAL') || die();
+
+class secret_manager {
+
+    const REVOKED = 'revoked';
+    const VALID = 'valid';
+    const NONVALID = 'nonvalid';
+
+    private $factor;
+    private $sessionvar;
+
+    public function __construct(string $factor) {
+        $this->factor = $factor;
+        $this->sessionvar = "tool_mfa_secrets_{$this->factor}";
+    }
+
+    /**
+     * This function creates or takes a secret, and stores it in the database or session.
+     *
+     * @param integer $expires the length of time the secret is valid. e.g. 1 min = 60
+     * @param boolean $session whether this secret should be stored in the session.
+     * @param string $secret an optional provided secret
+     * @return string the secret code, or 0 if no new code created.
+     */
+    public function create_secret(int $expires, bool $session, string $secret = null) : string {
+
+        // Check if there already an active secret, unless we are forcibly given a code.
+        if ($this->has_active_secret() && empty($secret)) {
+            return '';
+        }
+
+        // Setup a secret if not provided.
+        if (empty($secret)) {
+            $secret = random_int(100000, 999999);
+        }
+
+        // Now pass the code where it needs to go.
+        if ($session) {
+            $this->add_secret_to_session($secret, $expires);
+        } else {
+            $this->add_secret_to_db($secret, $expires);
+        }
+
+        return $secret;
+    }
+
+    /**
+     * Inserts the provided secret into the database with a given expiry duration.
+     *
+     * @param string $secret the secret to store
+     * @param integer $expires expiry duration in seconds
+     * @return void
+     */
+    private function add_secret_to_db(string $secret, int $expires) : void {
+        global $DB, $USER;
+        $expirytime = time() + $expires;
+
+        $data = [
+            'userid' => $USER->id,
+            'factor' => $this->factor,
+            'secret' => $secret,
+            'timecreated' => time(),
+            'expiry' => $expirytime,
+            'revoked' => 0
+        ];
+        $DB->insert_record('tool_mfa_secrets', $data);
+    }
+
+    /**
+     * Inserts the provided secret into the session with a given expiry duration.
+     *
+     * @param string $secret the secret to store
+     * @param integer $expires expiry duration in seconds
+     * @return void
+     */
+    private function add_secret_to_session(string $secret, int $expires) : void {
+        global $SESSION;
+
+        $expirytime = time() + $expires;
```

Comment:
So this is me being devils advocate and just asking

If we want a secret which is tied to the session, do we want to force this to only be available in the session (which is a different question)?


ie I'm thinking ahead to the situation where we have out of band secret approvals, we need to be able to approve a secret without being the active user and having access to the $SESSION, but in this case it still makes sense for the secret to be tied to a single session and not able to be used for other sessions.

In other words should all secrets be stored in the db, but they have an optional sessionid column which is checked as needed?